File: Android/src/build-backups/kapt3-20260412-000124/stubs/debug/selfgemma/talk/ui/llmchat/LlmChatContextManager.java
package selfgemma.talk.ui.llmchat;

@kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000H\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0000\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\b\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\r\n\u0002\u0018\u0002\n\u0002\b\u0002\b\u0000\u0018\u0000 #2\u00020\u0001:\u0001#B\u0011\u0012\b\b\u0002\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0004\b\u0004\u0010\u0005JF\u0010\u0006\u001a\u00020\u00072\u0006\u0010\b\u001a\u00020\t2\f\u0010\n\u001a\b\u0012\u0004\u0012\u00020\f0\u000b2\u0006\u0010\r\u001a\u00020\t2\u0006\u0010\u000e\u001a\u00020\u000f2\u0006\u0010\u0010\u001a\u00020\u000f2\u0006\u0010\u0011\u001a\u00020\u00122\b\b\u0002\u0010\u0013\u001a\u00020\u0014J$\u0010\u0015\u001a\b\u0012\u0004\u0012\u00020\t0\u000b2\f\u0010\u0016\u001a\b\u0012\u0004\u0012\u00020\t0\u000b2\u0006\u0010\u0017\u001a\u00020\u000fH\u0002J,\u0010\u0018\u001a\b\u0012\u0004\u0012\u00020\t0\u000b2\f\u0010\u0019\u001a\b\u0012\u0004\u0012\u00020\t0\u000b2\u0006\u0010\u0017\u001a\u00020\u000f2\u0006\u0010\u0013\u001a\u00020\u0014H\u0002J\u0018\u0010\u001a\u001a\u00020\t2\u0006\u0010\u001b\u001a\u00020\t2\u0006\u0010\u0017\u001a\u00020\u000fH\u0002J,\u0010\u001c\u001a\u00020\t2\u0006\u0010\b\u001a\u00020\t2\f\u0010\u001d\u001a\b\u0012\u0004\u0012\u00020\t0\u000b2\f\u0010\u001e\u001a\b\u0012\u0004\u0012\u00020\t0\u000bH\u0002J\u0012\u0010\u001f\u001a\u0004\u0018\u00010\t2\u0006\u0010 \u001a\u00020\fH\u0002J\f\u0010!\u001a\u00020\t*\u00020\"H\u0002R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006$"}, d2 = {"Lselfgemma/talk/ui/llmchat/LlmChatContextManager;", "", "tokenEstimator", "Lselfgemma/talk/domain/roleplay/usecase/TokenEstimator;", "<init>", "(Lselfgemma/talk/domain/roleplay/usecase/TokenEstimator;)V", "buildPlan", "Lselfgemma/talk/ui/llmchat/LlmChatContextPlan;", "baseSystemPrompt", "", "historyMessages", "", "Lselfgemma/talk/ui/common/chat/ChatMessage;", "pendingInput", "pendingImageCount", "", "pendingAudioCount", "contextProfile", "Lselfgemma/talk/domain/roleplay/model/ModelContextProfile;", "preferredMode", "Lselfgemma/talk/ui/llmchat/LlmChatContextMode;", "selectRecentLines", "historyLines", "budgetTokens", "buildSummaryLines", "olderLines", "fitToBudget", "text", "buildPrompt", "summaryLines", "recentLines", "toHistoryLine", "message", "toHistorySpeaker", "Lselfgemma/talk/ui/common/chat/ChatSide;", "Companion", "app_debug"})
public final class LlmChatContextManager {
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.TokenEstimator tokenEstimator = null;
    @org.jetbrains.annotations.NotNull()
    private static final kotlin.text.Regex WHITESPACE_REGEX = null;
    @org.jetbrains.annotations.NotNull()
    private static final selfgemma.talk.ui.llmchat.LlmChatContextManager.Companion Companion = null;
    
    public LlmChatContextManager(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.TokenEstimator tokenEstimator) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    public final selfgemma.talk.ui.llmchat.LlmChatContextPlan buildPlan(@org.jetbrains.annotations.NotNull()
    java.lang.String baseSystemPrompt, @org.jetbrains.annotations.NotNull()
    java.util.List<? extends selfgemma.talk.ui.common.chat.ChatMessage> historyMessages, @org.jetbrains.annotations.NotNull()
    java.lang.String pendingInput, int pendingImageCount, int pendingAudioCount, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.ModelContextProfile contextProfile, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.ui.llmchat.LlmChatContextMode preferredMode) {
        return null;
    }
    
    private final java.util.List<java.lang.String> selectRecentLines(java.util.List<java.lang.String> historyLines, int budgetTokens) {
        return null;
    }
    
    private final java.util.List<java.lang.String> buildSummaryLines(java.util.List<java.lang.String> olderLines, int budgetTokens, selfgemma.talk.ui.llmchat.LlmChatContextMode preferredMode) {
        return null;
    }
    
    private final java.lang.String fitToBudget(java.lang.String text, int budgetTokens) {
        return null;
    }
    
    private final java.lang.String buildPrompt(java.lang.String baseSystemPrompt, java.util.List<java.lang.String> summaryLines, java.util.List<java.lang.String> recentLines) {
        return null;
    }
    
    private final java.lang.String toHistoryLine(selfgemma.talk.ui.common.chat.ChatMessage message) {
        return null;
    }
    
    private final java.lang.String toHistorySpeaker(selfgemma.talk.ui.common.chat.ChatSide $this$toHistorySpeaker) {
        return null;
    }
    
    public LlmChatContextManager() {
        super();
    }
    
    @kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000\u0014\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\b\u0003\b\u0082\u0003\u0018\u00002\u00020\u0001B\t\b\u0002\u00a2\u0006\u0004\b\u0002\u0010\u0003R\u0011\u0010\u0004\u001a\u00020\u0005\u00a2\u0006\b\n\u0000\u001a\u0004\b\u0006\u0010\u0007\u00a8\u0006\b"}, d2 = {"Lselfgemma/talk/ui/llmchat/LlmChatContextManager$Companion;", "", "<init>", "()V", "WHITESPACE_REGEX", "Lkotlin/text/Regex;", "getWHITESPACE_REGEX", "()Lkotlin/text/Regex;", "app_debug"})
    static final class Companion {
        
        private Companion() {
            super();
        }
        
        @org.jetbrains.annotations.NotNull()
        public final kotlin.text.Regex getWHITESPACE_REGEX() {
            return null;
        }
    }
}

File: Android/src/app/src/main/java/com/google/ai/edge/gallery/ui/llmchat/LlmChatContextManager.kt
package selfgemma.talk.ui.llmchat

import selfgemma.talk.domain.roleplay.model.ModelContextProfile
import selfgemma.talk.domain.roleplay.usecase.TokenEstimator
import selfgemma.talk.ui.common.chat.ChatMessage
import selfgemma.talk.ui.common.chat.ChatMessageAudioClip
import selfgemma.talk.ui.common.chat.ChatMessageImage
import selfgemma.talk.ui.common.chat.ChatMessageText
import selfgemma.talk.ui.common.chat.ChatMessageType
import selfgemma.talk.ui.common.chat.ChatSide
import com.google.ai.edge.litertlm.Contents

private const val PER_IMAGE_TOKEN_RESERVE = 256
private const val PER_AUDIO_TOKEN_RESERVE = 192
private const val HISTORY_LINE_MAX_CHARS = 240
private const val SUMMARY_LINE_MAX_CHARS = 140

internal enum class LlmChatContextMode {
  FULL,
  AGGRESSIVE,
}

internal data class LlmChatContextReport(
  val usableInputTokens: Int,
  val reservedForCurrentTurnTokens: Int,
  val availableInstructionTokens: Int,
  val estimatedInstructionTokens: Int,
  val currentTurnOverflowDetected: Boolean,
  val mode: LlmChatContextMode,
  val recentLineCount: Int,
  val summaryLineCount: Int,
  val droppedLineCount: Int,
  val systemPromptTrimmed: Boolean,
)

internal data class LlmChatContextPlan(
  val systemInstruction: Contents?,
  val report: LlmChatContextReport,
)

internal class LlmChatContextManager(
  private val tokenEstimator: TokenEstimator = TokenEstimator(),
) {
  fun buildPlan(
    baseSystemPrompt: String,
    historyMessages: List<ChatMessage>,
    pendingInput: String,
    pendingImageCount: Int,
    pendingAudioCount: Int,
    contextProfile: ModelContextProfile,
    preferredMode: LlmChatContextMode = LlmChatContextMode.FULL,
  ): LlmChatContextPlan {
    val historyLines = historyMessages.mapNotNull(::toHistoryLine)
    val pendingInputTokens =
      tokenEstimator.estimate(pendingInput) +
        (pendingImageCount * PER_IMAGE_TOKEN_RESERVE) +
        (pendingAudioCount * PER_AUDIO_TOKEN_RESERVE)
    val currentTurnOverflowDetected = pendingInputTokens > contextProfile.usableInputTokens
    val availableInstructionTokens =
      (contextProfile.usableInputTokens - pendingInputTokens).coerceAtLeast(0)

    val promptBudget =
      when (preferredMode) {
        LlmChatContextMode.FULL -> availableInstructionTokens
        LlmChatContextMode.AGGRESSIVE -> availableInstructionTokens
      }
    val normalizedBasePrompt = baseSystemPrompt.trim()
    val basePrompt = fitToBudget(normalizedBasePrompt, budgetTokens = promptBudget / 2)
    val systemPromptTrimmed =
      normalizedBasePrompt.isNotBlank() &&
        tokenEstimator.estimate(basePrompt) < tokenEstimator.estimate(normalizedBasePrompt)
    var remainingBudget = (promptBudget - tokenEstimator.estimate(basePrompt)).coerceAtLeast(0)

    val recentBudgetShare =
      when (preferredMode) {
        LlmChatContextMode.FULL -> 0.65
        LlmChatContextMode.AGGRESSIVE -> 0.45
      }
    val recentBudget = (remainingBudget * recentBudgetShare).toInt().coerceAtLeast(0)
    val summaryBudget = (remainingBudget - recentBudget).coerceAtLeast(0)

    val recentLines = selectRecentLines(historyLines = historyLines, budgetTokens = recentBudget)
    val olderLines = historyLines.dropLast(recentLines.size)
    val summaryLines =
      buildSummaryLines(
        olderLines = olderLines,
        budgetTokens = summaryBudget,
        preferredMode = preferredMode,
      )

    val prompt =
      buildPrompt(
        baseSystemPrompt = basePrompt,
        summaryLines = summaryLines,
        recentLines = recentLines,
      )
    val fittedPrompt = fitToBudget(prompt, budgetTokens = promptBudget)
    val estimatedTokens = tokenEstimator.estimate(fittedPrompt)
    val droppedLineCount =
      (historyLines.size - recentLines.size - summaryLines.size).coerceAtLeast(0)

    return LlmChatContextPlan(
      systemInstruction = fittedPrompt.takeIf { it.isNotBlank() }?.let(Contents::of),
      report =
        LlmChatContextReport(
          usableInputTokens = contextProfile.usableInputTokens,
          reservedForCurrentTurnTokens = pendingInputTokens,
          availableInstructionTokens = availableInstructionTokens,
          estimatedInstructionTokens = estimatedTokens,
          currentTurnOverflowDetected = currentTurnOverflowDetected,
          mode =
            when {
              currentTurnOverflowDetected -> LlmChatContextMode.AGGRESSIVE
              droppedLineCount > 0 || summaryLines.isEmpty() && olderLines.isNotEmpty() ->
                LlmChatContextMode.AGGRESSIVE
              else -> preferredMode
            },
          recentLineCount = recentLines.size,
          summaryLineCount = summaryLines.size,
          droppedLineCount = droppedLineCount,
          systemPromptTrimmed = systemPromptTrimmed,
        ),
    )
  }

  private fun selectRecentLines(historyLines: List<String>, budgetTokens: Int): List<String> {
    if (historyLines.isEmpty() || budgetTokens <= 0) {
      return emptyList()
    }

    val selected = mutableListOf<String>()
    var usedTokens = 0
    for (line in historyLines.asReversed()) {
      val lineTokens = tokenEstimator.estimate(line)
      if (usedTokens + lineTokens > budgetTokens) {
        break
      }
      selected += line
      usedTokens += lineTokens
    }
    return selected.asReversed()
  }

  private fun buildSummaryLines(
    olderLines: List<String>,
    budgetTokens: Int,
    preferredMode: LlmChatContextMode,
  ): List<String> {
    if (olderLines.isEmpty() || budgetTokens <= 0) {
      return emptyList()
    }

    val cappedOlderLines =
      when (preferredMode) {
        LlmChatContextMode.FULL -> olderLines.takeLast(8)
        LlmChatContextMode.AGGRESSIVE -> olderLines.takeLast(4)
      }

    val summaryLines = mutableListOf<String>()
    var usedTokens = 0
    cappedOlderLines.forEach { line ->
      val summaryLine = "- ${line.take(SUMMARY_LINE_MAX_CHARS)}"
      val summaryTokens = tokenEstimator.estimate(summaryLine)
      if (usedTokens + summaryTokens > budgetTokens) {
        return@forEach
      }
      summaryLines += summaryLine
      usedTokens += summaryTokens
    }
    return summaryLines
  }

  private fun fitToBudget(text: String, budgetTokens: Int): String {
    if (text.isBlank() || budgetTokens <= 0) {
      return ""
    }
    val normalized = text.replace(WHITESPACE_REGEX, " ").trim()
    if (tokenEstimator.estimate(normalized) <= budgetTokens) {
      return normalized
    }
    val maxChars = (budgetTokens * 4).coerceAtLeast(0)
    if (maxChars == 0) {
      return ""
    }
    return normalized.take(maxChars).trim()
  }

  private fun buildPrompt(
    baseSystemPrompt: String,
    summaryLines: List<String>,
    recentLines: List<String>,
  ): String {
    return buildString {
      if (baseSystemPrompt.isNotBlank()) {
        appendLine(baseSystemPrompt)
      }
      if (summaryLines.isNotEmpty() || recentLines.isNotEmpty()) {
        if (isNotEmpty()) {
          appendLine()
        }
        appendLine("Use the following compressed conversation memory to stay consistent with the ongoing chat.")
      }
      if (summaryLines.isNotEmpty()) {
        appendLine()
        appendLine("Earlier turns summary:")
        summaryLines.forEach(::appendLine)
      }
      if (recentLines.isNotEmpty()) {
        appendLine()
        appendLine("Recent turns:")
        recentLines.forEach(::appendLine)
      }
    }
      .trim()
  }

  private fun toHistoryLine(message: ChatMessage): String? {
    return when (message) {
      is ChatMessageText ->
        message.content
          .trim()
          .takeIf { it.isNotBlank() }
          ?.replace(WHITESPACE_REGEX, " ")
          ?.take(HISTORY_LINE_MAX_CHARS)
          ?.let { "${message.side.toHistorySpeaker()}: $it" }
      is ChatMessageImage ->
        when (message.side) {
          ChatSide.USER -> "User shared ${message.bitmaps.size} image(s)."
          ChatSide.AGENT -> "Assistant shared image output."
          ChatSide.SYSTEM -> null
        }
      is ChatMessageAudioClip ->
        when (message.side) {
          ChatSide.USER -> "User shared an audio clip."
          ChatSide.AGENT -> "Assistant shared audio output."
          ChatSide.SYSTEM -> null
        }
      else ->
        when (message.type) {
          ChatMessageType.LOADING,
          ChatMessageType.ERROR,
          ChatMessageType.WARNING,
          ChatMessageType.THINKING,
          ChatMessageType.PROMPT_TEMPLATES,
          ChatMessageType.CONFIG_VALUES_CHANGE,
          ChatMessageType.COLLAPSABLE_PROGRESS_PANEL,
          ChatMessageType.BENCHMARK_RESULT,
          ChatMessageType.BENCHMARK_LLM_RESULT,
          ChatMessageType.CLASSIFICATION,
          ChatMessageType.WEBVIEW,
          ChatMessageType.IMAGE_WITH_HISTORY,
          ChatMessageType.INFO -> null
          ChatMessageType.TEXT,
          ChatMessageType.IMAGE,
          ChatMessageType.AUDIO_CLIP -> null
        }
    }
  }

  private fun ChatSide.toHistorySpeaker(): String {
    return when (this) {
      ChatSide.USER -> "User"
      ChatSide.AGENT -> "Assistant"
      ChatSide.SYSTEM -> "System"
    }
  }

  private companion object {
    val WHITESPACE_REGEX = Regex("\\s+")
  }
}
